File: algorithms/sorting/cycle-sort/python/cycle_sort.py
def cycle_sort(arr: list[int]) -> list[int]:
    """
    Cycle Sort implementation.
    An in-place, unstable sorting algorithm that is optimal in terms of
    the number of writes to the original array.
    """
    result = list(arr)
    n = len(result)

    # Traverse array elements to find where they belong
    for cycle_start in range(0, n - 1):
        item = result[cycle_start]

        # Find position where we put the item
        pos = cycle_start
        for i in range(cycle_start + 1, n):
            if result[i] < item:
                pos += 1

        # If item is already in correct position
        if pos == cycle_start:
            continue

        # Ignore all duplicate elements
        while item == result[pos]:
            pos += 1

        # Put the item to its right position
        result[pos], item = item, result[pos]

        # Rotate the rest of the cycle
        while pos != cycle_start:
            pos = cycle_start
            for i in range(cycle_start + 1, n):
                if result[i] < item:
                    pos += 1

            while item == result[pos]:
                pos += 1

            result[pos], item = item, result[pos]

    return result

File: algorithms/sorting/cycle-sort/go/CycleSort.go
package main

import "fmt"

func CycleSort(arr []int) []int {
	n := len(arr)

	for cycleStart := 0; cycleStart < n-1; cycleStart++ {
		item := arr[cycleStart]

		// Find the position where we put the item
		pos := cycleStart
		for i := cycleStart + 1; i < n; i++ {
			if arr[i] < item {
				pos++
			}
		}

		// If the item is already in the correct position
		if pos == cycleStart {
			continue
		}

		// Skip duplicates
		for item == arr[pos] {
			pos++
		}

		// Put the item to its correct position
		if pos != cycleStart {
			item, arr[pos] = arr[pos], item
		}

		// Rotate the rest of the cycle
		for pos != cycleStart {
			pos = cycleStart

			for i := cycleStart + 1; i < n; i++ {
				if arr[i] < item {
					pos++
				}
			}

			for item == arr[pos] {
				pos++
			}

			if item != arr[pos] {
				item, arr[pos] = arr[pos], item
			}
		}
	}

	return arr
}

func main() {
	arr := []int{5, 3, 8, 1, 2, -3, 0}
	fmt.Println(CycleSort(arr))
}
